### ISSUE TITLE
Can ranger automatically generate thumbnails for all files for preview in a directory?

### ISSUE THREAD

**Issue by hleft** [ranger/ranger#2784]
Some gui file managers (like dolphin) can automatically generate thumbnails for all files for preview, is there a way for ranger to do this?
 
Because some file previews are a bit slow so I would like to be able to autogenerate before reaching that file

**Comment by markus-bauer** [comment]
Depending on what you're asking, there might be issues about this already:
1) preview a directory (or directory content) as if it was a file (in the right panel) :
https://github.com/ranger/ranger/pull/2157
2) show thumnbails in the main panel, like a gui:
This probably won't happen, but there's an issue about it:
https://github.com/ranger/ranger/issues/2224

**Comment by hleft** [comment]
@markus-bauer 

It doesn't look the same. I just want to automatically generate thumbnails of all the files to speed up the preview, not to change the ui or the preview directory.

**Comment by markus-bauer** [comment]
@hleft Ah, sorry. You mean you want to trigger the generating of previews for all or some files in the current directory, without scrolling over the files first.

**Comment by hleft** [comment]
@markus-bauer  Yes hopefully this can be done simply. (Of course, it is good to add a thumbnail mode to the ui, but that may be a complicated requirement)

**Comment by markus-bauer** [comment]
I don't know of a good way to do it currently, but perhaps someone else can think of something.

As a hack, you can manually trigger the creation of all previews with a custom command (in $CONFIG/ranger/commands.py).

```python
class preload_previews(Command):
    def execute(self):
        # Previews need to know the size of the preview area, so we have to get that first.
        # pager is either the pager, or the preview column.
        pager = self.fm.ui.get_pager()
        if pager:
            (wid, hei) = (pager.wid, pager.hei)
            # generate previews for all files in this dir (for this wid, hei)
            # could be modified to just use current selection.
            for file in self.fm.thisdir.files:
                if file.is_file:
                    self.fm.notify(f"preloading preview for: {file.path}")
                    # This will generate the preview.
                    file.get_preview_source(wid, hei)
```

I think it works. It does create preview tasks for all files (see taskview). But, again, perhaps someone else knows a better solution.

**Comment by hleft** [comment]
@markus-bauer Thank you very much this is what I need

**Comment by hleft** [comment]
@markus-bauer One more question, how to further recursively generate previews for all subdirectories?

**Comment by markus-bauer** [comment]
Sorry, I don't have time at the moment. I think doing this for all subdirs might be more complicated. I don't know if I can come up with something on the spot.
You can just call :flat first, before preload_previews.

**Comment by toonn** [comment]
This code is a good candidate for `commands.py` IMO.
